draft-ietf-quic-version-negotiation.md

```diff
@@ -210,6 +189,54 @@ balancers to ensure that packets for a given connection are routed to the same
 server.
 
 
+# Server Deployments of QUIC {#server-fleet}
+
+While this document mainly discusses a single QUIC server, it is common for
+deployments of QUIC servers to include a fleet of multiple server instances. We
+therefore define the following terms:
+
+Supported Versions:
+
+: This is the set of versions supported by a given individual server instance.
+
+Fully-Deployed Versions:
+
+: This is the set of QUIC versions that is supported by every single QUIC
+server instance in this deployment.
+
+Partially-Deployed Versions:
+
+: This is the set of versions supported by a given individual server instance,
+but which aren't supported by every single QUIC server instance in this
+deployment.
+
+If a deployment only contains a single server instance, then the Supported
+Versions is equal to the Fully-Deployed Versions, and the Partially-Deployed
+Versions is the empty set.
+
+Conversely, if a deployment contains multiple server instances, it is possible
+for software updates to not happen at the exact same time on all server
+instances. In this scenario, some instances will have a different Supported
+Versions from other instances in the same fleet. When that happens, a client
+might receive a Version Negotiation packet from a server instance that doesn't
+support some Partially-Supported Versions; in that case, those
+Partially-Supported Versions won't be listed in the Version Negotiation packet.
+When the client reacts to that Version Negotiation packet, it might create a
+new QUIC connection which could be routed to a different server instance, which
+has a different set of Supported Versions. The version downgrade prevention
+mechanism described in {{downgrade}} is robust to this scenario and will not
+falsely detect a downgrade attack.
+
+If a server operator is progressively deploying a new QUIC version throughout
+its fleet, it MUST perform a two-step process where it first progressively adds
+support for the new version. This adds the new version to Supported Versions
+but not to Fully-Deployed Versions. Once all server instances have been
+upgraded, the second step is to progressively add the version to Fully-Deployed
+Versions on all server instances. This opens connections to version downgrades
```

Comment:
This needs to be three steps:

1. Add code for version X
2. Advertise version X in Version Negotiation
3. Add version X to Other Versions

The reason I say this is the in-flight VN.  Let's say that you support version X and you just add it both VN and OV.  What about clients that 2ms ago got a VN without X in it?

The transition between step 2 and 3 can be short.  You can run that on a fairly short timer for a single server deployment, but you need to ensure that all VNs are out of the system before you move to step 3 or connections will fail.

The logic here is:

* don't update the advisory signals (VN) until the code is in place
* don't update the downgrade protection (OV) until all the advisory signals are in place

Rollback is similar: first remove downgrade protection, then remove advisory signals, then remove actual support.

One thing I like about this design is that we because we don't rely on VN being correct, the design can be made robust against VN being wrong for a short period.

See also https://quicwg.org/ops-drafts/draft-ietf-quic-applicability.html#name-enabling-new-versions